File: skeleton/src/main/java/com/example/routing/api/rest/ErrorResponse.java
package com.example.routing.api.rest;

import com.fasterxml.jackson.annotation.JsonInclude;

import java.time.Instant;
import java.util.List;

@JsonInclude(JsonInclude.Include.NON_NULL)
public record ErrorResponse(
    Instant timestamp,
    int status,
    String errorCode,
    String message,
    String path,
    String traceId,
    List<FieldError> errors
) {
    public record FieldError(
        String field,
        Object rejectedValue,
        String message
    ) {}

    public static ErrorResponse of(int status, String errorCode, String message, String path) {
        String traceId = org.slf4j.MDC.get("traceId");
        return new ErrorResponse(Instant.now(), status, errorCode, message, path, traceId, null);
    }

    public static ErrorResponse withFieldErrors(int status, String errorCode, String message,
                                                String path, List<FieldError> errors) {
        String traceId = org.slf4j.MDC.get("traceId");
        return new ErrorResponse(Instant.now(), status, errorCode, message, path, traceId, errors);
    }
}

File: pilot_app/lib/core/network/error_response.dart
/// Resposta de erro do backend (4xx/5xx). Doc 14 checklist.
class ErrorResponse {
  const ErrorResponse({
    required this.timestamp,
    required this.status,
    this.errorCode,
    this.message,
    this.path,
    this.traceId,
  });

  final DateTime timestamp;
  final int status;
  final String? errorCode;
  final String? message;
  final String? path;
  final String? traceId;

  factory ErrorResponse.fromJson(Map<String, dynamic> json) {
    return ErrorResponse(
      timestamp: json['timestamp'] != null
          ? DateTime.tryParse(json['timestamp'] as String) ?? DateTime.now()
          : DateTime.now(),
      status: json['status'] as int? ?? 0,
      errorCode: json['errorCode'] as String?,
      message: json['message'] as String?,
      path: json['path'] as String?,
      traceId: json['traceId'] as String?,
    );
  }
}
